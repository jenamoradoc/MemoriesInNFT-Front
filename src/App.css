.container {
  display: grid;
  grid-template-columns: 400px calc(100% - 400px);
  align-items: center;
}
