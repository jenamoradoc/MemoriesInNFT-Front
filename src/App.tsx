import { FC, useCallback, useState } from "react";

import { Results, Search } from "./components";
import { LoginResults } from "./interfaces";
import "./App.css";

const App: FC = () => {
  const [queryLogin, setQueryLogin] = useState<string>("");
  const [results, setResults] = useState<LoginResults[]>([]);
  const [loading, setLoading] = useState<boolean>(false);
  const [message, setMessage] = useState<string>("");

  const getLoginQuery = useCallback(
    (login: string) => setQueryLogin(login),
    []
  );

  const getResults = useCallback(
    (data: LoginResults[]) => setResults(data),
    []
  );

  const changeLoading = useCallback((state: boolean) => setLoading(state), []);

  const changeMessage = useCallback(
    (message: string) => setMessage(message),
    []
  );

  return (
    <div className="container">
      <Search
        changeLoading={changeLoading}
        changeMessage={changeMessage}
        getResults={getResults}
        loading={loading}
        message={message}
        results={results}
        usernameEntered={getLoginQuery}
      />
      <Results
        changeLoading={changeLoading}
        changeMessage={changeMessage}
        getResults={getResults}
        login={queryLogin}
        results={results}
      />
    </div>
  );
};

export default App;
