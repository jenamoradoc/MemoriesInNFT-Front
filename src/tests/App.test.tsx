import {
  render,
  screen,
  waitFor,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import App from "../App";

describe("Testing <App />", () => {
  test("Should render the initial content", () => {
    const { container } = render(<App />);

    expect(container).toMatchSnapshot();
    expect(container).toBeInTheDocument();
  });

  test("should show aside container", () => {
    render(<App />);

    expect(screen.getByRole("complementary")).toBeInTheDocument();
  });

  test("should show the input field", () => {
    render(<App />);

    expect(screen.getByRole("textbox")).toBeInTheDocument();
  });

  test("should be empty the input field", () => {
    render(<App />);

    expect((screen.getByRole("textbox") as HTMLInputElement).value.trim()).toBe(
      ""
    );
  });

  test("should show a Submit button", () => {
    render(<App />);
    expect(screen.getByText(/submit/i)).toBeInTheDocument();
  });

  test("should change the input value", async () => {
    render(<App />);

    userEvent.type(screen.getByRole("textbox"), "pedro");

    expect((screen.getByRole("textbox") as HTMLInputElement).value).toBe(
      "pedro"
    );
  });

  test("should show results when the input field is changed and the button is clicked", async () => {
    const { container } = render(<App />);

    userEvent.type(screen.getByRole("textbox"), "pedro");
    userEvent.click(screen.getByText(/submit/i));

    expect(screen.getByTestId("spinner")).toBeInTheDocument();

    await waitForElementToBeRemoved(() => screen.queryByTestId("spinner"));

    await screen.findByRole("list");
    expect(screen.getAllByRole("listitem").length).toBeGreaterThan(0);

    expect(container).toMatchSnapshot();
  });

  test("should display the previous and next page buttons", async () => {
    render(<App />);

    userEvent.type(screen.getByRole("textbox"), "pedro");
    userEvent.click(screen.getByText(/submit/i));

    expect(screen.getByTestId("spinner")).toBeInTheDocument();

    await waitFor(() => screen.findByRole("list"));
    expect(screen.getAllByRole("listitem").length).toBeGreaterThan(0);

    expect(screen.getByText("«")).toBeTruthy();
    expect(screen.getByText("»")).toBeTruthy();
  });

  test("should show the next results when the Next Page button is clicked", async () => {
    render(<App />);

    userEvent.type(screen.getByRole("textbox"), "pedro");
    userEvent.click(screen.getByText(/submit/i));

    expect(screen.getByTestId("spinner")).toBeInTheDocument();

    await waitForElementToBeRemoved(() => screen.queryByTestId("spinner"));

    if (
      screen.getAllByRole("listitem").length === 0 &&
      screen.getByText(/something went wrong/i)
    ) {
      return;
    }

    await screen.findByRole("list");
    expect(screen.getAllByRole("listitem").length).toBeGreaterThan(0);

    userEvent.click(screen.getByText("»"));
    expect(screen.getByTestId("spinner")).toBeInTheDocument();
    await waitForElementToBeRemoved(screen.queryByTestId("spinner"));

    await waitFor(() => screen.findByRole("list"));
    expect(screen.getAllByRole("listitem").length).toBeGreaterThan(0);
  });

  test("should show the no results page when there are no results in the search", async () => {
    render(<App />);

    userEvent.type(screen.getByRole("textbox"), "wejhifhejwfiwehfuewhuw");
    userEvent.click(screen.getByText(/submit/i));

    expect(screen.getByTestId("spinner")).toBeInTheDocument();

    await waitForElementToBeRemoved(() => screen.queryByTestId("spinner"));

    expect(screen.getByText(/results not found/i)).toBeInTheDocument();
  });

  test("should clear the input element when the Clear Filter button is clicked", async () => {
    render(<App />);

    userEvent.type(screen.getByRole("textbox"), "pedro");
    userEvent.click(screen.getByText(/submit/i));

    userEvent.click(screen.getByText(/clear filter/i));

    expect((screen.getByRole("textbox") as HTMLInputElement).value.trim()).toBe(
      ""
    );
  });
});
