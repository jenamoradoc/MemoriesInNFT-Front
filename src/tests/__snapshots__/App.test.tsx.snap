// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <App /> Should render the initial content 1`] = `
<div>
  <div
    class="container"
  >
    <aside
      class="search__form-container"
    >
      <form
        class="search__form"
      >
        <h1
          class="search__form-title"
        >
          Search github user(s)
        </h1>
        <input
          class="search__form-input"
          type="text"
          value=""
        />
        <div
          class="search__form-btns__container"
        >
          <button
            class="search__form-btn"
            type="submit"
          >
            Submit
          </button>
          <button
            class="search__form-btn"
          >
            Clear Filter
          </button>
        </div>
      </form>
    </aside>
    <div>
      <div
        class="results__message"
      >
        <h2>
          GitHub User Search App
        </h2>
        <img
          alt="Team"
          draggable="false"
          src="team.svg"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Testing <App /> should show results when the input field is changed and the button is clicked 1`] = `
<div>
  <div
    class="container"
  >
    <aside
      class="search__form-container"
    >
      <form
        class="search__form"
      >
        <h1
          class="search__form-title"
        >
          Search github user(s)
        </h1>
        <input
          class="search__form-input"
          type="text"
          value="pedro"
        />
        <div
          class="search__form-btns__container"
        >
          <button
            class="search__form-btn"
            type="submit"
          >
            Submit
          </button>
          <button
            class="search__form-btn"
          >
            Clear Filter
          </button>
        </div>
      </form>
    </aside>
    <div>
      <select
        class="results__sort"
      >
        <option
          disabled=""
          value=""
        >
          --Choose an option
        </option>
        <option
          value="best match"
        >
          Best Match
        </option>
        <option
          value="followers"
        >
          Followers
        </option>
        <option
          value="repositories"
        >
          Repositories
        </option>
        <option
          value="joined"
        >
          Joined
        </option>
      </select>
      <ul
        class="results__list"
      >
        <li
          class="results__list-item"
        >
          <img
            alt="pedro"
            class="results__list-item-image"
            draggable="false"
            src="https://avatars.githubusercontent.com/u/2248?v=4"
          />
          <strong
            class="results__list-item-login"
          >
            pedro
          </strong>
          <span
            class="results__list-item-type"
          >
            User
          </span>
        </li>
        <li
          class="results__list-item"
        >
          <img
            alt="pedrovgs"
            class="results__list-item-image"
            draggable="false"
            src="https://avatars.githubusercontent.com/u/4030704?v=4"
          />
          <strong
            class="results__list-item-login"
          >
            pedrovgs
          </strong>
          <span
            class="results__list-item-type"
          >
            User
          </span>
        </li>
        <li
          class="results__list-item"
        >
          <img
            alt="pedronauck"
            class="results__list-item-image"
            draggable="false"
            src="https://avatars.githubusercontent.com/u/2029172?v=4"
          />
          <strong
            class="results__list-item-login"
          >
            pedronauck
          </strong>
          <span
            class="results__list-item-type"
          >
            User
          </span>
        </li>
        <li
          class="results__list-item"
        >
          <img
            alt="pedromassango"
            class="results__list-item-image"
            draggable="false"
            src="https://avatars.githubusercontent.com/u/33294549?v=4"
          />
          <strong
            class="results__list-item-login"
          >
            pedromassango
          </strong>
          <span
            class="results__list-item-type"
          >
            User
          </span>
        </li>
        <li
          class="results__list-item"
        >
          <img
            alt="pedrouid"
            class="results__list-item-image"
            draggable="false"
            src="https://avatars.githubusercontent.com/u/10136079?v=4"
          />
          <strong
            class="results__list-item-login"
          >
            pedrouid
          </strong>
          <span
            class="results__list-item-type"
          >
            User
          </span>
        </li>
        <li
          class="results__list-item"
        >
          <img
            alt="Pedromdsn"
            class="results__list-item-image"
            draggable="false"
            src="https://avatars.githubusercontent.com/u/57453260?v=4"
          />
          <strong
            class="results__list-item-login"
          >
            Pedromdsn
          </strong>
          <span
            class="results__list-item-type"
          >
            User
          </span>
        </li>
        <li
          class="results__list-item"
        >
          <img
            alt="pedroslopez"
            class="results__list-item-image"
            draggable="false"
            src="https://avatars.githubusercontent.com/u/4368928?v=4"
          />
          <strong
            class="results__list-item-login"
          >
            pedroslopez
          </strong>
          <span
            class="results__list-item-type"
          >
            User
          </span>
        </li>
        <li
          class="results__list-item"
        >
          <img
            alt="pedromarins"
            class="results__list-item-image"
            draggable="false"
            src="https://avatars.githubusercontent.com/u/215004?v=4"
          />
          <strong
            class="results__list-item-login"
          >
            pedromarins
          </strong>
          <span
            class="results__list-item-type"
          >
            User
          </span>
        </li>
        <li
          class="results__list-item"
        >
          <img
            alt="pedrualves"
            class="results__list-item-image"
            draggable="false"
            src="https://avatars.githubusercontent.com/u/1184178?v=4"
          />
          <strong
            class="results__list-item-login"
          >
            pedrualves
          </strong>
          <span
            class="results__list-item-type"
          >
            User
          </span>
        </li>
      </ul>
      <div
        class="buttons__container"
      >
        <button
          disabled=""
          style="border-radius: 5px 0 0 5px;"
        >
          «
        </button>
        <button
          data-testid="1"
          disabled=""
        >
          1
        </button>
        <button
          data-testid="2"
        >
          2
        </button>
        <button
          data-testid="3"
        >
          3
        </button>
        <button
          data-testid="4"
        >
          4
        </button>
        <button
          data-testid="5"
        >
          5
        </button>
        <button
          data-testid="6"
        >
          6
        </button>
        <button
          data-testid="7"
        >
          7
        </button>
        <button
          data-testid="8"
        >
          8
        </button>
        <button
          data-testid="9"
        >
          9
        </button>
        <button
          style="border-radius: 0 5px 5px 0;"
        >
          »
        </button>
      </div>
    </div>
  </div>
</div>
`;
